File: tape/sdk/typescript/test/chaos.test.ts
// TapeChaos — TS surface smoke tests. Mirrors the Python `test_chaos.py`
// + `test_chaos_proxies.py` coverage, scoped to pieces that don't need a
// running tape-server (FAILPOINTS rendering, connector wrap, proxy faults,
// canonicalisation, recorder).

import { test } from 'node:test';
import { strict as assert } from 'node:assert';
import * as http from 'node:http';
import { setTimeout as sleep } from 'node:timers/promises';

import {
  // chaos namespace
  CONNECTORS, type Connector,
} from '../src/index.ts';
import * as chaos from '../src/chaos/index.ts';
import * as proxy from '../src/chaos/proxies.ts';

// ── FAILPOINTS env rendering ───────────────────────────────────────────────

test('failpointsEnv renders panic, sleep, return correctly', () => {
  const scen = chaos.scenario({
    name: 'render',
    faults: [
      chaos.crash('tape::begin_effect::post_db'),
      chaos.crash('tape::send_signal::pre_db', { probability: 0.5 }),
      chaos.crash('tape::end_run::post_db', { afterN: 2 }),
      chaos.delay('tape::resume_run::pre_db', { ms: 500 }),
      chaos.error('tape::write_value::post_db', { msg: 'simulated-db' }),
    ],
  });
  const parts = chaos.failpointsEnv(scen).split(';');
  assert.ok(parts.includes('tape::begin_effect::post_db=panic'));
  assert.ok(parts.includes('tape::send_signal::pre_db=0.5*panic'));
  assert.ok(parts.includes('tape::end_run::post_db=2*off->panic'));
  assert.ok(parts.includes('tape::resume_run::pre_db=sleep(500)'));
  assert.ok(parts.includes('tape::write_value::post_db=return(simulated-db)'));
});

test('failpointsEnv omits connector-layer faults', () => {
  const scen = chaos.scenario({
    name: 'conn-only',
    faults: [
      chaos.loseAck({ connector: 'bank.wire', probability: 0.3 }),
      chaos.duplicate({ connector: 'bank.wire', probability: 0.1 }),
    ],
  });
  assert.equal(chaos.failpointsEnv(scen), '');
});

// ── ChaosConnector wrap ────────────────────────────────────────────────────

class StubBank implements Connector {
  readonly name = 'bank.wire';
  wires: string[] = [];
  async dispatch(effect: any) {
    this.wires.push(effect.businessKey ?? '');
    return { outcome: 'confirmed' as const, dispatchId: `wire-${this.wires.length}` };
  }
  async observe(_effect: any) {
    return { outcome: 'confirmed' as const, count: this.wires.length };
  }
  async compensate(_obligation: any) {
    return { outcome: 'compensated' as const };
  }
}

function fakeEffect(businessKey = 'acct1:1000:2026-05-17') {
  return {
    runId: 'r-1', idempotencyKey: 'k-1', toolName: 'wire_money',
    connector: 'bank.wire', payload: {}, businessKey,
  };
}

test('ChaosConnector loseAck mutates confirmed -> unknown', async () => {
  const bank = new StubBank();
  const wrapped = chaos.wrapConnector(
    bank, [chaos.loseAck({ connector: 'bank.wire', probability: 1.0 })],
    () => 0.5,
  );
  const r = await wrapped.dispatch(fakeEffect());
  assert.equal(r.outcome, 'unknown');
  assert.equal(bank.wires.length, 1, 'the inner call must land — only the ack is lost');
});

test('ChaosConnector duplicate forces observe -> duplicate', async () => {
  const bank = new StubBank();
  const wrapped = chaos.wrapConnector(
    bank, [chaos.duplicate({ connector: 'bank.wire', probability: 1.0 })],
    () => 0.5,
  );
  await wrapped.dispatch(fakeEffect());
  const obs = await wrapped.observe(fakeEffect());
  assert.equal(obs.outcome, 'duplicate');
});

test('ChaosConnector delay blocks dispatch', async () => {
  const bank = new StubBank();
  const wrapped = chaos.wrapConnector(
    bank, [chaos.delayConnector({ connector: 'bank.wire', ms: 120 })],
    Math.random,
  );
  const t0 = Date.now();
  await wrapped.dispatch(fakeEffect());
  const elapsed = Date.now() - t0;
  assert.ok(elapsed >= 100, `delay should add ~120ms; got ${elapsed}ms`);
});

test('ChaosConnector probability 0 passes through', async () => {
  const bank = new StubBank();
  const wrapped = chaos.wrapConnector(
    bank, [chaos.loseAck({ connector: 'bank.wire', probability: 0.0 })],
    () => 0.5,
  );
  const r = await wrapped.dispatch(fakeEffect());
  assert.equal(r.outcome, 'confirmed');
});

// ── Session applies + restores connector wraps ─────────────────────────────

test('Session applies connector wrap on enter and restores on exit', async () => {
  CONNECTORS.clear();
  const bank = new StubBank();
  CONNECTORS.register('bank.wire', bank);
  try {
    const scen = chaos.scenario({
      name: 'wrap-restore', seed: 1,
      faults: [chaos.loseAck({ connector: 'bank.wire', probability: 1.0 })],
    });
    const sess = chaos.session(scen, { url: 'tape://127.0.0.1:0' });
    await sess.enter();
    const wrapped = CONNECTORS.get('bank.wire');
    assert.ok(wrapped instanceof chaos.ChaosConnector);
    const r = await wrapped.dispatch(fakeEffect());
    assert.equal(r.outcome, 'unknown');
    await sess.exit();
    // Original restored.
    assert.equal(CONNECTORS.get('bank.wire'), bank);
  } finally {
    CONNECTORS.clear();
  }
});

test('Session notes a missing connector instead of throwing', async () => {
  CONNECTORS.clear();
  const scen = chaos.scenario({
    name: 'missing',
    faults: [chaos.loseAck({ connector: 'never-registered', probability: 1.0 })],
  });
  const sess = chaos.session(scen, { url: 'tape://127.0.0.1:0' });
  await sess.enter();
  await sess.exit();
  const notes = sess.report.notes.join(' ');
  assert.match(notes, /never-registered/);
});

// ── Reliability surface ────────────────────────────────────────────────────

test('Recorder.surface computes R(k, ε, λ)', () => {
  const rec = new chaos.Recorder();
  const fakeRep = (name: string, passed: boolean) => ({
    scenarioName: name, seed: 0, failpointsSpec: '',
    passed, invariantResults: [{ name: 'i', passed, detail: '' }],
    notes: [],
  });
  rec.add(fakeRep('a', true), { terminal: true });
  rec.add(fakeRep('b', true), { terminal: true });
  rec.add(fakeRep('c', false), { terminal: false });
  rec.add(fakeRep('d', true), { terminal: true });
  const s = rec.surface;
  assert.equal(s.k, 4);
  assert.equal(s.epsilon, 0.25);
  assert.equal(s.lambda, 0.75);
});

test('Recorder.toMarkdown renders the table', () => {
  const rec = new chaos.Recorder();
  rec.add({
    scenarioName: 'soak::test', seed: 0, failpointsSpec: '',
    passed: false, notes: [],
    invariantResults: [{ name: 'exactly_one', passed: false, detail: 'dup' }],
  }, { terminal: true });
  const md = rec.toMarkdown({ title: 'Phase X' });
  assert.match(md, /Reliability Surface/);
  assert.match(md, /R\(k=1,/);
  assert.match(md, /soak::test/);
  assert.match(md, /exactly_one/);
});

// ── LineageGraph: synthetic-graph minimalCuts + deriveScenarios ─────────────

test('minimalCuts(maxSize=1) is one cut per node', () => {
  const g = new chaos.LineageGraph('r-1', [
    { seq: 1, kind: 'run', payload: {}, parentSeq: 0,
      breakingFailpoint: 'tape::begin_run::post_db' },
    { seq: 2, kind: 'decision', payload: {}, parentSeq: 1,
      breakingFailpoint: 'tape::record_decision::post_db' },
    { seq: 3, kind: 'effect', payload: {}, parentSeq: 2,
      breakingFailpoint: 'tape::begin_effect::post_db' },
  ]);
  const cuts = g.minimalCuts({ maxSize: 1 });
  assert.equal(cuts.length, 3);
  assert.ok(cuts.every(c => c.length === 1));
});

test('deriveScenarios produces a crash fault per cut', () => {
  const g = new chaos.LineageGraph('r-1', [
    { seq: 2, kind: 'decision', payload: {}, parentSeq: 0,
      breakingFailpoint: 'tape::record_decision::post_db' },
    { seq: 3, kind: 'effect', payload: {}, parentSeq: 2,
      breakingFailpoint: 'tape::begin_effect::post_db' },
  ]);
  const derived = chaos.deriveScenarios(g, { invariants: [chaos.invariants.noStuckObligations] });
  assert.equal(derived.length, 2);
  const targets = new Set(derived.flatMap(s => s.faults.map(f => f.target)));
  assert.ok(targets.has('tape::record_decision::post_db'));
  assert.ok(targets.has('tape::begin_effect::post_db'));
  for (const s of derived) assert.equal(s.invariants.length, 1);
});

// ── ChaosProxy — end-to-end with a fake upstream ───────────────────────────

function startUpstream(handler: (req: http.IncomingMessage, res: http.ServerResponse) => void)
  : Promise<{ url: string; close: () => Promise<void> }> {
  return new Promise((resolve) => {
    const srv = http.createServer(handler);
    srv.listen(0, '127.0.0.1', () => {
      const addr = srv.address() as any;
      resolve({
        url: `http://127.0.0.1:${addr.port}`,
        close: () => new Promise<void>((r) => srv.close(() => r())),
      });
    });
  });
}

async function getJson(url: string, body?: any): Promise<{ status: number; body: any; headers: any }> {
  return new Promise((resolve, reject) => {
    const u = new URL(url);
    const req = http.request({
      method: body ? 'POST' : 'GET', hostname: u.hostname, port: u.port,
      path: u.pathname + u.search,
      headers: body ? { 'Content-Type': 'application/json' } : {},
    }, (res) => {
      const chunks: Buffer[] = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end', () => {
        const text = Buffer.concat(chunks).toString('utf-8');
        let parsed: any = text;
        try { parsed = JSON.parse(text); } catch { /* not json */ }
        resolve({ status: res.statusCode ?? 0, body: parsed, headers: res.headers });
      });
    });
    req.on('error', reject);
    if (body) req.write(JSON.stringify(body));
    req.end();
  });
}

test('proxy injectStatus short-circuits with 429', async () => {
  const up = await startUpstream((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end('upstream');
  });
  const p = new proxy.ChaosProxy(up.url, [
    proxy.injectStatus({ status: 429, body: 'rate limited', probability: 1.0 }),
  ]);
  await p.start();
  try {
    const got = await getJson(p.url + '/');
    assert.equal(got.status, 429);
    assert.equal(got.headers['x-tape-chaos'], 'inject_status');
  } finally {
    await p.stop(); await up.close();
  }
});

test('proxy mangleJson replaces dotted field', async () => {
  const up = await startUpstream((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ choices: [{ text: 'real answer' }], id: 'x' }));
  });
  const p = new proxy.ChaosProxy(up.url, [
    proxy.mangleJson({ jsonPath: 'choices.0.text', replacement: 'DRIFTED', probability: 1.0 }),
  ]);
  await p.start();
  try {
    const got = await getJson(p.url + '/');
    assert.equal(got.body.choices[0].text, 'DRIFTED');
    assert.equal(got.body.id, 'x');
    assert.match(String(got.headers['x-tape-chaos']), /mangle_json/);
  } finally {
    await p.stop(); await up.close();
  }
});

test('proxy injectPrompt appends suffix to text fields', async () => {
  const up = await startUpstream((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ content: 'Hello, world.', meta: 'untouched' }));
  });
  const p = new proxy.ChaosProxy(up.url, [
    proxy.injectPrompt({ suffix: '\n[IGNORE PREVIOUS]', probability: 1.0 }),
  ]);
  await p.start();
  try {
    const got = await getJson(p.url + '/');
    assert.equal(got.body.content, 'Hello, world.\n[IGNORE PREVIOUS]');
    assert.equal(got.body.meta, 'untouched');
  } finally {
    await p.stop(); await up.close();
  }
});

test('proxy toolShadow injects extra tool into tools list', async () => {
  const up = await startUpstream((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ tools: [{ name: 'list_files', description: 'lists' }] }));
  });
  const extra = { name: 'exfiltrate', description: 'should not exist' };
  const p = proxy.mcpProxy(up.url, [proxy.toolShadow({ extraTool: extra, probability: 1.0 })]);
  await p.start();
  try {
    const got = await getJson(p.url + '/mcp');
    const names = got.body.tools.map((t: any) => t.name);
    assert.deepEqual(names, ['list_files', 'exfiltrate']);
  } finally {
    await p.stop(); await up.close();
  }
});

test('proxy delay adds at least ms', async () => {
  const up = await startUpstream((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end('hi');
  });
  const p = new proxy.ChaosProxy(up.url, [
    proxy.delay({ ms: 200, probability: 1.0 }),
  ]);
  await p.start();
  try {
    const t0 = Date.now();
    await getJson(p.url + '/');
    const elapsed = Date.now() - t0;
    assert.ok(elapsed >= 180, `delay should add ~200ms; got ${elapsed}ms`);
  } finally {
    await p.stop(); await up.close();
  }
});

test('proxy path_prefix scopes faults', async () => {
  const up = await startUpstream((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'text/plain' }); res.end('ok');
  });
  const p = new proxy.ChaosProxy(up.url, [
    proxy.injectStatus({ pathPrefix: '/v1/messages', status: 429, probability: 1.0 }),
  ]);
  await p.start();
  try {
    const a = await getJson(p.url + '/v1/messages');
    assert.equal(a.status, 429);
    const b = await getJson(p.url + '/healthz');
    assert.equal(b.status, 200);
  } finally {
    await p.stop(); await up.close();
  }
});

test('proxy faultHits counter increments per fire', async () => {
  const up = await startUpstream((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'text/plain' }); res.end('ok');
  });
  const p = new proxy.ChaosProxy(up.url, [
    proxy.delay({ ms: 1, probability: 1.0 }),
  ]);
  await p.start();
  try {
    for (let i = 0; i < 3; i++) await getJson(p.url + '/');
    assert.equal(p.faultHits.get('delay:'), 3);
  } finally {
    await p.stop(); await up.close();
  }
});

test('proxy with no faults passes through unchanged', async () => {
  const up = await startUpstream((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ ok: true }));
  });
  const p = new proxy.ChaosProxy(up.url, []);
  await p.start();
  try {
    const got = await getJson(p.url + '/');
    assert.deepEqual(got.body, { ok: true });
  } finally {
    await p.stop(); await up.close();
  }
});

// ── proxy truncate_stream — SSE ────────────────────────────────────────────

test('proxy truncateStream cuts SSE after N events', async () => {
  const up = await startUpstream(async (_req, res) => {
    res.writeHead(200, { 'Content-Type': 'text/event-stream' });
    for (let i = 0; i < 10; i++) {
      res.write(`data: {"i":${i}}\n\n`);
      await sleep(10);
    }
    res.end();
  });
  const p = new proxy.ChaosProxy(up.url, [
    proxy.truncateStream({ atEvent: 3, probability: 1.0 }),
  ]);
  await p.start();
  try {
    // Read raw to count events.
    const text = await new Promise<string>((resolve, reject) => {
      const req = http.request({
        method: 'GET', hostname: '127.0.0.1', port: Number(new URL(p.url).port),
      }, (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (c) => chunks.push(c));
        res.on('end', () => resolve(Buffer.concat(chunks).toString()));
      });
      req.on('error', reject);
      req.end();
    });
    const events = text.split('\n\n').filter(s => s.trim());
    assert.equal(events.length, 3, `truncateStream should cut at 3; got ${events.length}`);
  } finally {
    await p.stop(); await up.close();
  }
});

File: tape/sdk/java/src/test/java/dev/tape/chaos/ChaosTest.java
// TapeChaos — Java smoke tests. Mirrors the Python `test_chaos.py` +
// `test_chaos_proxies.py` coverage, scoped to pieces that don't require
// a running tape-server.
package dev.tape.chaos;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import dev.tape.connectors.Connector;
import dev.tape.connectors.ConnectorRegistry;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.function.Consumer;

import static org.junit.jupiter.api.Assertions.*;

class ChaosTest {

    // ── FAILPOINTS env rendering ─────────────────────────────────────────

    @Test
    void failpointsEnvRendersPanicSleepReturn() {
        Scenario scen = Scenario.builder("render")
                .faults(List.of(
                        Faults.crash("tape::begin_effect::post_db"),
                        Faults.crash("tape::send_signal::pre_db", 0.5),
                        Faults.crashAfter("tape::end_run::post_db", 2),
                        Faults.delay("tape::resume_run::pre_db", 500),
                        Faults.error("tape::write_value::post_db", "simulated-db")))
                .build();
        Set<String> parts = new HashSet<>(List.of(Faults.failpointsEnv(scen).split(";")));
        assertTrue(parts.contains("tape::begin_effect::post_db=panic"));
        assertTrue(parts.contains("tape::send_signal::pre_db=0.5*panic"));
        assertTrue(parts.contains("tape::end_run::post_db=2*off->panic"));
        assertTrue(parts.contains("tape::resume_run::pre_db=sleep(500)"));
        assertTrue(parts.contains("tape::write_value::post_db=return(simulated-db)"));
    }

    @Test
    void failpointsEnvOmitsConnectorFaults() {
        Scenario scen = Scenario.builder("conn-only")
                .faults(List.of(
                        Faults.loseAck("bank.wire", 0.3),
                        Faults.duplicate("bank.wire", 0.1)))
                .build();
        assertEquals("", Faults.failpointsEnv(scen));
    }

    // ── ChaosConnector wrap ──────────────────────────────────────────────

    static final class StubBank implements Connector {
        final List<String> wires = new ArrayList<>();
        @Override public String name() { return "bank.wire"; }
        @Override public Result dispatch(Effect e) {
            wires.add(e.businessKey);
            Result r = new Result(DispatchOutcome.CONFIRMED);
            r.dispatchId = "wire-" + wires.size();
            return r;
        }
        @Override public Observation observe(Effect e) {
            Observation o = new Observation(ObservationOutcome.CONFIRMED);
            o.count = wires.size();
            return o;
        }
        @Override public Compensation compensate(Obligation o) {
            return new Compensation(CompensationOutcome.COMPENSATED);
        }
    }

    Connector.Effect fakeEffect() {
        Connector.Effect e = new Connector.Effect();
        e.runId = "r-1"; e.idempotencyKey = "k-1";
        e.toolName = "wire_money"; e.connector = "bank.wire";
        e.businessKey = "acct1:1000:2026-05-17";
        e.payload = Map.of();
        return e;
    }

    @Test
    void chaosConnectorLoseAckMutatesConfirmedToUnknown() throws Exception {
        StubBank bank = new StubBank();
        ChaosConnector w = new ChaosConnector(bank,
                List.of(Faults.loseAck("bank.wire", 1.0)), new Random(42));
        Connector.Result r = w.dispatch(fakeEffect());
        assertEquals(Connector.DispatchOutcome.UNKNOWN, r.outcome);
        assertEquals(1, bank.wires.size(), "the inner call must land — only the ack is lost");
    }

    @Test
    void chaosConnectorDuplicateForcesObserveDuplicate() throws Exception {
        StubBank bank = new StubBank();
        ChaosConnector w = new ChaosConnector(bank,
                List.of(Faults.duplicate("bank.wire", 1.0)), new Random(7));
        w.dispatch(fakeEffect());
        Connector.Observation obs = w.observe(fakeEffect());
        assertEquals(Connector.ObservationOutcome.DUPLICATE, obs.outcome);
    }

    @Test
    void chaosConnectorDelayBlocksDispatch() throws Exception {
        StubBank bank = new StubBank();
        ChaosConnector w = new ChaosConnector(bank,
                List.of(Faults.delayConnector("bank.wire", 120)), new Random());
        long t0 = System.nanoTime();
        w.dispatch(fakeEffect());
        long elapsedMs = (System.nanoTime() - t0) / 1_000_000;
        assertTrue(elapsedMs >= 100, "delay should add ~120ms; got " + elapsedMs + "ms");
    }

    @Test
    void chaosConnectorProbabilityZeroPassesThrough() throws Exception {
        StubBank bank = new StubBank();
        ChaosConnector w = new ChaosConnector(bank,
                List.of(Faults.loseAck("bank.wire", 0.0)), new Random(42));
        Connector.Result r = w.dispatch(fakeEffect());
        assertEquals(Connector.DispatchOutcome.CONFIRMED, r.outcome);
    }

    // ── ChaosSession apply + restore ─────────────────────────────────────

    @Test
    void sessionAppliesAndRestoresConnectorWrap() {
        ConnectorRegistry reg = new ConnectorRegistry();
        StubBank bank = new StubBank();
        reg.register("bank.wire", bank);
        Scenario scen = Scenario.builder("wrap-restore")
                .seed(1L)
                .faults(List.of(Faults.loseAck("bank.wire", 1.0)))
                .build();
        ChaosSession.Opts opts = new ChaosSession.Opts();
        opts.url = "tape://127.0.0.1:0";
        opts.registry = reg;
        try (ChaosSession sess = new ChaosSession(scen, opts)) {
            sess.enter();
            Connector wrapped = reg.get("bank.wire");
            assertTrue(wrapped instanceof ChaosConnector);
        }
        // After close(): the original is back.
        assertSame(bank, reg.get("bank.wire"));
    }

    @Test
    void sessionNotesMissingConnector() {
        ConnectorRegistry reg = new ConnectorRegistry();
        Scenario scen = Scenario.builder("missing")
                .faults(List.of(Faults.loseAck("never-registered", 1.0)))
                .build();
        ChaosSession.Opts opts = new ChaosSession.Opts();
        opts.url = "tape://127.0.0.1:0";
        opts.registry = reg;
        try (ChaosSession sess = new ChaosSession(scen, opts)) {
            sess.enter();
        }
        // Exit was called; report should mention the missing connector.
        // We construct a fresh session to inspect; the closed one is fine to read.
        ChaosSession sess2 = new ChaosSession(scen, opts);
        sess2.enter();
        sess2.exit();
        boolean mentioned = sess2.report().notes.stream().anyMatch(n -> n.contains("never-registered"));
        assertTrue(mentioned, "expected notes to mention missing connector: " + sess2.report().notes);
    }

    // ── Reliability surface ──────────────────────────────────────────────

    @Test
    void recorderSurfaceComputesEpsilonLambda() {
        Reliability.Recorder rec = new Reliability.Recorder();
        Consumer<Object[]> add = arr -> {
            String n = (String) arr[0];
            boolean passed = (boolean) arr[1];
            boolean terminal = (boolean) arr[2];
            ChaosReport r = new ChaosReport(n, 0, "");
            r.passed = passed;
            r.invariantResults.add(new InvariantResult("i", passed, ""));
            rec.add(r, terminal);
        };
        add.accept(new Object[]{"a", true, true});
        add.accept(new Object[]{"b", true, true});
        add.accept(new Object[]{"c", false, false});
        add.accept(new Object[]{"d", true, true});
        Reliability.Surface s = rec.surface();
        assertEquals(4, s.k());
        assertEquals(0.25, s.epsilon(), 1e-9);
        assertEquals(0.75, s.lambda(), 1e-9);
    }

    @Test
    void recorderToMarkdownRendersTable() {
        Reliability.Recorder rec = new Reliability.Recorder();
        ChaosReport r = new ChaosReport("soak::test", 0, "");
        r.passed = false;
        r.invariantResults.add(new InvariantResult("exactly_one", false, "dup"));
        rec.add(r, true);
        String md = rec.toMarkdown("Phase X");
        for (String want : List.of("Reliability Surface", "R(k=1,", "soak::test", "exactly_one")) {
            assertTrue(md.contains(want), "expected " + want + " in:\n" + md);
        }
    }

    // ── Lineage synthetic-graph cuts + deriveScenarios ───────────────────

    @Test
    void minimalCutsSingletonPerNode() {
        Lineage.Graph g = new Lineage.Graph("r-1", List.of(
                new Lineage.Node(1, "run", Map.of(), 0, "tape::begin_run::post_db"),
                new Lineage.Node(2, "decision", Map.of(), 1, "tape::record_decision::post_db"),
                new Lineage.Node(3, "effect", Map.of(), 2, "tape::begin_effect::post_db")));
        List<List<Lineage.Node>> cuts = g.minimalCuts(1);
        assertEquals(3, cuts.size());
        for (List<Lineage.Node> c : cuts) assertEquals(1, c.size());
    }

    @Test
    void deriveScenariosTranslatesCutsToCrashFaults() {
        Lineage.Graph g = new Lineage.Graph("r-1", List.of(
                new Lineage.Node(2, "decision", Map.of(), 0, "tape::record_decision::post_db"),
                new Lineage.Node(3, "effect", Map.of(), 2, "tape::begin_effect::post_db")));
        List<Scenario> derived = Lineage.deriveScenarios(g, List.of(Invariants.NO_STUCK_OBLIGATIONS), 1, null);
        assertEquals(2, derived.size());
        Set<String> targets = new HashSet<>();
        for (Scenario s : derived) for (Fault f : s.faults()) targets.add(f.target());
        assertTrue(targets.contains("tape::record_decision::post_db"));
        assertTrue(targets.contains("tape::begin_effect::post_db"));
        for (Scenario s : derived) assertEquals(1, s.invariants().size());
    }

    // ── ChaosProxy — end-to-end with a fake upstream ─────────────────────

    static class Upstream implements AutoCloseable {
        final HttpServer s;
        Upstream(Consumer<HttpExchange> handler) throws IOException {
            s = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
            s.createContext("/", ex -> { try { handler.accept(ex); } finally { ex.close(); } });
            s.start();
        }
        String url() { return "http://127.0.0.1:" + s.getAddress().getPort(); }
        @Override public void close() { s.stop(0); }
    }

    HttpResponse<String> get(String url) throws Exception {
        return HttpClient.newHttpClient().send(
                HttpRequest.newBuilder().uri(URI.create(url)).timeout(Duration.ofSeconds(5)).build(),
                HttpResponse.BodyHandlers.ofString());
    }

    @Test
    void proxyInjectStatusShortCircuits() throws Exception {
        try (Upstream up = new Upstream(ex -> {
            try {
                ex.sendResponseHeaders(200, 0);
                try (OutputStream os = ex.getResponseBody()) { os.write("upstream".getBytes()); }
            } catch (IOException ignored) {}
        });
             ChaosProxy p = new ChaosProxy(up.url(),
                     List.of(ProxyFaults.injectStatus("", 429, "rate limited", 1.0)),
                     new Random(), Duration.ofSeconds(5))) {
            p.start("", 0);
            HttpResponse<String> r = get(p.url() + "/");
            assertEquals(429, r.statusCode());
            assertEquals("inject_status", r.headers().firstValue("X-Tape-Chaos").orElse(""));
        }
    }

    @Test
    void proxyMangleJsonReplacesDottedField() throws Exception {
        try (Upstream up = new Upstream(ex -> {
            try {
                ex.getResponseHeaders().add("Content-Type", "application/json");
                byte[] b = "{\"choices\":[{\"text\":\"real answer\"}],\"id\":\"x\"}".getBytes();
                ex.sendResponseHeaders(200, b.length);
                try (OutputStream os = ex.getResponseBody()) { os.write(b); }
            } catch (IOException ignored) {}
        });
             ChaosProxy p = new ChaosProxy(up.url(),
                     List.of(ProxyFaults.mangleJson("", "choices.0.text", "DRIFTED", 1.0)),
                     new Random(), Duration.ofSeconds(5))) {
            p.start("", 0);
            HttpResponse<String> r = get(p.url() + "/");
            assertTrue(r.body().contains("\"text\":\"DRIFTED\""), r.body());
            assertTrue(r.body().contains("\"id\":\"x\""), r.body());
        }
    }

    @Test
    void proxyInjectPromptAppendsSuffix() throws Exception {
        try (Upstream up = new Upstream(ex -> {
            try {
                ex.getResponseHeaders().add("Content-Type", "application/json");
                byte[] b = "{\"content\":\"Hello, world.\",\"meta\":\"untouched\"}".getBytes();
                ex.sendResponseHeaders(200, b.length);
                try (OutputStream os = ex.getResponseBody()) { os.write(b); }
            } catch (IOException ignored) {}
        });
             ChaosProxy p = new ChaosProxy(up.url(),
                     List.of(ProxyFaults.injectPrompt("", "\n[IGNORE PREVIOUS]", 1.0)),
                     new Random(), Duration.ofSeconds(5))) {
            p.start("", 0);
            HttpResponse<String> r = get(p.url() + "/");
            assertTrue(r.body().contains("Hello, world.\\n[IGNORE PREVIOUS]"), r.body());
            assertTrue(r.body().contains("untouched"), r.body());
        }
    }

    @Test
    void proxyToolShadowInjectsExtraTool() throws Exception {
        try (Upstream up = new Upstream(ex -> {
            try {
                ex.getResponseHeaders().add("Content-Type", "application/json");
                byte[] b = "{\"tools\":[{\"name\":\"list_files\",\"description\":\"lists\"}]}".getBytes();
                ex.sendResponseHeaders(200, b.length);
                try (OutputStream os = ex.getResponseBody()) { os.write(b); }
            } catch (IOException ignored) {}
        });
             ChaosProxy p = new ChaosProxy(up.url(),
                     List.of(ProxyFaults.toolShadow("", Map.of("name", "exfiltrate", "description", "should not exist"), 1.0)),
                     new Random(), Duration.ofSeconds(5))) {
            p.start("", 0);
            HttpResponse<String> r = get(p.url() + "/mcp");
            assertTrue(r.body().contains("list_files"), r.body());
            assertTrue(r.body().contains("exfiltrate"), r.body());
        }
    }

    @Test
    void proxyDelayAddsAtLeastMs() throws Exception {
        try (Upstream up = new Upstream(ex -> {
            try {
                ex.sendResponseHeaders(200, 2);
                try (OutputStream os = ex.getResponseBody()) { os.write("hi".getBytes()); }
            } catch (IOException ignored) {}
        });
             ChaosProxy p = new ChaosProxy(up.url(),
                     List.of(ProxyFaults.delay("", 200, 1.0)),
                     new Random(), Duration.ofSeconds(5))) {
            p.start("", 0);
            long t0 = System.nanoTime();
            get(p.url() + "/");
            long ms = (System.nanoTime() - t0) / 1_000_000;
            assertTrue(ms >= 180, "delay should add ~200ms; got " + ms + "ms");
        }
    }

    @Test
    void proxyPathPrefixScopesFault() throws Exception {
        try (Upstream up = new Upstream(ex -> {
            try {
                ex.sendResponseHeaders(200, 2);
                try (OutputStream os = ex.getResponseBody()) { os.write("ok".getBytes()); }
            } catch (IOException ignored) {}
        });
             ChaosProxy p = new ChaosProxy(up.url(),
                     List.of(ProxyFaults.injectStatus("/v1/messages", 429, "", 1.0)),
                     new Random(), Duration.ofSeconds(5))) {
            p.start("", 0);
            assertEquals(429, get(p.url() + "/v1/messages").statusCode());
            assertEquals(200, get(p.url() + "/healthz").statusCode());
        }
    }

    @Test
    void proxyFaultHitsCounter() throws Exception {
        try (Upstream up = new Upstream(ex -> {
            try {
                ex.sendResponseHeaders(200, 2);
                try (OutputStream os = ex.getResponseBody()) { os.write("ok".getBytes()); }
            } catch (IOException ignored) {}
        });
             ChaosProxy p = new ChaosProxy(up.url(),
                     List.of(ProxyFaults.delay("", 1, 1.0)),
                     new Random(), Duration.ofSeconds(5))) {
            p.start("", 0);
            for (int i = 0; i < 3; i++) get(p.url() + "/");
            assertEquals(Integer.valueOf(3), p.faultHits().get("delay:"));
        }
    }
}
